### ISSUE TITLE
** (MatchError) no match of right hand side value: {:error, {:error, :invalid}}

### ISSUE THREAD

**Issue by rawkode** [florinpatrascu/neo4j_sips#20]
No matter how I configure this library, I can't seem to execute any queries. I keep getting:

```elixir
** (MatchError) no match of right hand side value: {:error, {:error, :invalid}}
    (neo4j_sips) lib/neo4j_sips/query.ex:12: Neo4j.Sips.Query.query/2
    (neo4j_sips) lib/neo4j_sips/query.ex:30: Neo4j.Sips.Query.query!/2
```

`conn` gives me
```elixir
iex(1)> Neo4j.Sips.conn()
%Neo4j.Sips.Connection{commit_url: "", options: nil,
 server: %Neo4j.Sips.Server{bolt: nil,
  data: %{batch: "http://neo4j:7474/db/data/batch",
    constraints: "http://neo4j:7474/db/data/schema/constraint",
    cypher: "http://neo4j:7474/db/data/cypher", extensions: %{},
    extensions_info: "http://neo4j:7474/db/data/ext",
    indexes: "http://neo4j:7474/db/data/schema/index", neo4j_version: "3.1.3",
    node: "http://neo4j:7474/db/data/node",
    node_index: "http://neo4j:7474/db/data/index/node",
    node_labels: "http://neo4j:7474/db/data/labels",
    relationship: "http://neo4j:7474/db/data/relationship",
    relationship_index: "http://neo4j:7474/db/data/index/relationship",
    relationship_types: "http://neo4j:7474/db/data/relationship/types",
    transaction: "http://neo4j:7474/db/data/transaction"},
  data_url: "http://neo4j:7474/db/data/",
  management_url: "http://neo4j:7474/db/manage/",
  server_url: "http://neo4j:7474", timeout: 5000}, server_version: "3.1.3",
 transaction_url: "http://neo4j:7474/db/data/transaction"}
```

I've configured multiple ways:
```elixir
supervisor(Neo4j.Sips, [[url: "http://neo4j:7474", basic_auth: [username: "neo4j", password: "super-secret"]]]),
```
and through `config.exs`

```elixir
config :neo4j_sips, Neo4j,
  url: "http://neo4j:super-secret@neo4j:7474", 
  pool_size: 5,
  max_overflow: 2,
  timeout: 30
```

**Comment by florinpatrascu** [comment]
can you run the tests, locally?

**Comment by rawkode** [comment]
After changing https://github.com/florinpatrascu/neo4j_sips/blob/master/config/test.exs#L4
to use port 7474, yeah.

I can login to the web interface with my credentials.

**Comment by rawkode** [comment]
My code is

```elixir
Neo4j.Sips.query(Neo4j.Sips.conn, ~s/CREATE (n:List {name: "#{event.payload[:name]}"})/)
```

Neo4J version is 3.1

**Comment by florinpatrascu** [comment]
is this: `No matter how I configure this library, I can't seem to execute any queries` still an issue?

**Comment by rawkode** [comment]
@florinpatrascu Yes

**Comment by florinpatrascu** [comment]
Not sure what is going on there. I just tried successfully the following easy test:

```elixir
Interactive Elixir (1.4.2) - press Ctrl+C to exit (type h() ENTER for help)

iex(1)>  alias Neo4j.Sips, as: Neo4j
Neo4j.Sips

iex(2)> Neo4j.start_link(url: "http://localhost:7373")
{:ok, #PID<0.210.0>}

iex(3)>  Neo4j.query(Neo4j.conn, "return 1 as n")

12:08:05.562 [debug] POST - http://localhost:7373/db/data/transaction/commit - {"statements":[{"statement":"return 1 as n"}]}
{:ok, [%{"n" => 1}]}

iex(4)> Neo4j.server_version
"3.1.0"
```

My test Neo4j db is on port `7373`, no auth, fwiw. HTH

**Comment by florinpatrascu** [comment]
Also:

```elixir

iex(7)> Neo4j.query(Neo4j.conn, "CREATE (n:List {name: {name}})", %{name: "a_name"})

12:16:07.870 [debug] POST - http://localhost:7373/db/data/transaction/commit - {"statements":[{"statement":"CREATE (n:List {name: {name}})","parameters":{"name":"a_name"}}]}
{:ok, []}

iex(8)> Neo4j.query(Neo4j.conn, "match (n:List {name: 'a_name'}) return n")

12:17:38.523 [debug] POST - http://localhost:7373/db/data/transaction/commit - {"statements":[{"statement":"match (n:List {name: 'a_name'}) return n"}]}

{:ok, [%{"n" => %{"name" => "a_name"}}]}
```

**Comment by rawkode** [comment]
This must be something to-do with my setup. I'll tear it down and back up and we'll close this for now. Thanks for taking the time to help me out!

**Comment by florinpatrascu** [comment]
np, please let me know how it goes. I am planning anyway, to update the dependencies for this lib; clean up the code a bit, etc., and I can debug this further, if need be?! If you're planning to use Neo4j for a project, I'd kindly suggest to use `bolt_sips`. Not sure if the HTTP REST API will be supported for too long, by Neo4j. Bolt is definitely the way to go. HTH

**Comment by rawkode** [comment]
Yeah, I'll get Bolt working. Why "Sips" btw? I've seen a few Elixir projects use that suffix, I don't think I know what it means.

Have you ever used Janus graph? I'm torn between that and Neo4j

**Comment by florinpatrascu** [comment]
Sips? Hehe. Personally, when I first started this, I never imagined I could end up taking larger gulps from Neo4j, using Elixir; Sips == humble_beginnings, maybe?! :) But then slowly-slowly, I acquired the taste for this combo; Neo4j & Elixir. One can always use `alias`, when typing `Neo4j.Sips` is too much. Plus, I suck at naming products in English, named methods, etc.

Janus? Yeah, a very cool library indeed, but I can't stand Gremlin. Neo4j (Cypher) is really what works for me, and for others too, and since Bolt (w/ Elixir, or course) ...  unparalleled speed.
